src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/insights/client/BitbucketClient.java

```diff
@@ -0,0 +1,149 @@
+import com.github.mc1arke.sonarqube.plugin.ce.pullrequest.bitbucket.BitbucketServerPullRequestDecorator;
+import com.github.mc1arke.sonarqube.plugin.ce.pullrequest.bitbucket.insights.client.model.CreateAnnotationsRequest;
+import com.github.mc1arke.sonarqube.plugin.ce.pullrequest.bitbucket.insights.client.model.CreateReportRequest;
+import com.github.mc1arke.sonarqube.plugin.ce.pullrequest.bitbucket.insights.client.model.ServerProperties;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import org.sonar.api.ce.ComputeEngineSide;
+import org.sonar.api.config.Configuration;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import static java.lang.String.format;
+
+@ComputeEngineSide
+public class BitbucketClient {
+    private static final Logger LOGGER = Loggers.get(BitbucketClient.class);
+    private static final String REPORT_KEY = "com.github.mc1arke.sonarqube";
+    private final Configuration configuration;
+
+    private OkHttpClient client;
+    private ObjectMapper objectMapper;
+
+    public BitbucketClient(Configuration configuration) {
+        this.configuration = configuration;
+    }
+
+    public boolean isConfigured() {
+        return configuration.hasKey(BitbucketServerPullRequestDecorator.PULL_REQUEST_BITBUCKET_URL) &&
+                configuration.hasKey(BitbucketServerPullRequestDecorator.PULL_REQUEST_BITBUCKET_TOKEN);
+    }
+
+    public ServerProperties getServerProperties() throws IOException {
+        Request req = new Request.Builder()
+                .get()
+                .url(format("%s/rest/api/1.0/application-properties", baseUrl()))
+                .build();
+        try(Response response = getClient().newCall(req).execute()) {
+            validate(response);
+
+            return getObjectMapper().reader().forType(ServerProperties.class)
+                    .readValue(response.body().string());
+        }
+    }
+
+    public void createReport(String project, String repository, String commit, CreateReportRequest request) throws IOException {
+        String body = getObjectMapper().writeValueAsString(request);
+        Request req = new Request.Builder()
+                .put(RequestBody.create(MediaType.parse("application/json"), body))
+                .url(format("%s/rest/insights/1.0/projects/%s/repos/%s/commits/%s/reports/%s", baseUrl(), project, repository, commit, REPORT_KEY))
+                .build();
+
+        try (Response response = getClient().newCall(req).execute()) {
+            validate(response);
+        }
+    }
+
+    public void createAnnotations(String project, String repository, String commit, CreateAnnotationsRequest request) throws IOException {
+        if (request.getAnnotations().isEmpty()) {
+            return;
+        }
+        Request req = new Request.Builder()
```

Comment:
According to documentation:

> A report cannot have more than 1000 annotations by default, however this property is congurable at an instance level. If the request would result in more than the maximum number of annotations being stored then the entire request is rejected and no new annotations are stored.

We should make sure that for bigger projects all the annotations are still shown by splitting it up into max 1000 annotation batches.

Reply:
I have added logic for splitting the annotations in batches of 1000.
In addition I sort the issues by severity so if there are more than the max number of allowed annotations the ones with highest severity will be shown.